### PR TITLE
stats/distributions: make rv_sample public, allow subclassing

### LINKED ISSUE #8057: Subclass of scipy.stats.rv_discrete initiated with values throws an AttributeError on pmf method
Subclass of `scipy.stats.rv_discrete` initiated with values (i.e `rv_sample`) throws an error if `pmf `method is called. The same without subclassing works fine.

### Reproducing code example:
```
from scipy import stats 

class subc_rv_discrete(stats.rv_discrete):
    pass

xk = (0,1,2,3)
pk = (0.1, 0.2, 0.3, 0.4)

inst = subc_rv_discrete(values=(xk, pk))
inst.pmf(xk)
```
### Error message:
```
Traceback (most recent call last):

  File "<ipython-input-48-129655c38e6a>", line 10, in <module>
    inst.pmf(xk)

  File "C:\Anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 2832, in pmf
    args, loc, _ = self._parse_args(*args, **kwds)

AttributeError: 'rv_sample' object has no attribute '_parse_args'
```
### Scipy/Numpy/Python version information:
```
0.19.0 1.12.1 sys.version_info(major=3, minor=6, micro=1, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

closes gh-8057

#### What does this implement/fix?

Expose `rv_sample` to the main stats namespace. Make it play ball with subclassing: previously, subclassing `rv_discrete(values=(..., ...))` was broken. With this PR, subclassing `rv_sample` just works.

#### Additional information

After several failed attempts to fix the `__new__` dispatch in `rv_discrete` I'm throwing a towel on that. The dispatch to a subclass is a hack which apparently breaks all sorts of invariants. Which is, after all, expected --- this was originally devised as a hack around the `rv_discrete` quirk which was doing very different things in its `__init__`, depending on whether `values` are provided or not.

So, this PR instead introduces a new `rv_sample` subclass of `rv_discrete`. It is constructed with `xk, pk` arguments directly --- thus sidestepping the dispatch on `values=...`. I think it actually makes the signature lighter and more reasonable, too :-).

Deprecating `rv_discrete(values='...)` is out of scope of this PR.
Whether to actually deprecate it or not in a follow-up PR depends a bit on the timeline for the plans for revamping the distributions infrastructure. If it's planned in a finite amount of time from now, maybe there is no point causing downstream churn.

### CODE REVIEW

**Comment by mdhaber** [comment]
I don't think we need to deprecate the old use of `rv_discrete` for this purpose. I do think we need this to address the issue mentioned [here](https://github.com/scipy/scipy/pull/17236) now that we're making this public. I would do it by adding a lower and upper acceptable limit for each element of the support and inequality comparing against those (instead of strict equality comparisin with elements of the support) to determine whether the argument of a method is within the support. The limits can be below and above the elements of the support by a relative tolerance only (e.g. by multiplying each element of the support by `1-eps` and `1+eps`); I don't think an absolute tolerance is needed. If any of the intervals are overlapping, we could just raise an error when the `rv_sample` is created.

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:3824]: I hope this won't bother anyone, but it is a backward incompatibility that could be an issue if others looked for `rv_sample` as we did above. We would need to email the mailing list about it and list it in the release notes as a backward incompatibility. Alternatively, we could leave this as `rv_sample` and use a new name for the new class.

**Review by mdhaber — COMMENTED** [review]
Looks pretty good. Just a few questions and a comment about the test.

I wanted to explore whether we could just expose the existing `rv_sample` class. It looks like we could - what do you think of something like mdhaber/scipy#89? It passed all tests locally except for something in `..\..\scipy\stats\tests\test_distributions.py:88: in test_distributions_submodule` which looks easily fixable, and it probably just has to do with the fact that I made `rv_sample` public via `distributions.py`. (_Update: oops. It doesn't work when I try to freeze a subclass instance._)

Otherwise, the only comment is that we should allow for arguments to be "close" to elements in the domain or we'll get complaints, I think.

File: scipy/stats/_distn_infrastructure.py
Comment (ev-br) [reply scipy/stats/_distn_infrastructure.py:3824]: I'm not sure what is the backcompat issue? `rv_discrete(values=...)` keeps working unchanged, `isinstance(obj, rv_discrete)` keeps working. The specific class name did change, yes, so `obj.__class__.__name__` is different; my understanding is that `isinstance` checks are vastly preferred to explicit `__class__` checks, so I don't think we need to worry about these. What would be other backcompat issues?

That said, if you feel strongly about this, I'm ready to rename it back and use for the public class whatever name you come up with.

**Comment by ev-br** [comment]
> explore whether we could just expose the existing rv_sample class. It looks like we could .... (Update: oops. It doesn't work when I try to freeze a subclass instance.)

Exactly.
I had a few attempts, without success. IIUC, the hack where `rv_discrete.__new__` returns an `_rv_sample` instance breaks the subclassing chain. Somehow attempting to instantiate a subclass of `_rv_sample` gives back an `_rv_sample` instance, not a subclass instance. 

> Otherwise, the only comment is that we should allow for arguments to be "close" to elements in the domain or we'll get complaints, I think.

I'm not sure how you want it to play out, would you expand? Also whether you want it in the same PR or in a follow-up ( would that be a change in behavior?)

File: scipy/stats/_distn_infrastructure.py
Comment (ev-br) [reply scipy/stats/_distn_infrastructure.py:4012]: https://github.com/scipy/scipy/blob/main/scipy/stats/_distn_infrastructure.py#L440

It is basically used for picking/unpickling, accounting for potential  mutability of `__init__` parameters. IIRC, this is random_state, maybe `a` and `b`. It is full well possible that this is not needed across the board anymore with improvements over last couple of years.

File: scipy/stats/_distn_infrastructure.py
Comment (ev-br) [reply scipy/stats/_distn_infrastructure.py:3946]: `self.inc = 1` is not OK :-). This attribute is broken across all discrete distributions, the assumption is baked rather deep down. I once dropped into the test loop in `test_discrete_basic` a subclass with `self.inc=2` and a bunch of tests pass because in many places it really looks into `n+1` or some such.

That said, here this attribute is because of a test in `test_moments` which compares moments to a generic implementation IIRC, and that expects `self.inc` because of 
https://github.com/scipy/scipy/blob/main/scipy/stats/_distn_infrastructure.py#L3750

I did not want to touch tests in the PR (they could be improved, I'm sure)

File: scipy/stats/tests/test_discrete_distns.py
Comment (ev-br) [reply scipy/stats/tests/test_discrete_distns.py:576]: Facepalm. Will lump together with other tweaks.

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distn_infrastructure.py:3824]: I think I wasn't thinking straight. `rv_sample` wasn't public before, so there cannot have been an `isinstance` BC issue, and I agree that we don't need to be BC w.r.t. `obj.__class__.__name__`.

Nonetheless, it brings up the idea of choosing a more appropriate name for the class. Is "rv_sample" appropriate? That suggests to me that it is defined by a "sample" as `rv_histogram` is defined by a histogram, but I don't think that would be an accurate interpretation.  TBH `rv_discrete_histogram` would be apt, in a sense, but it may be confusing because the output of `np.histogram` cannot be used directly to instantiate it. `rv_count` or `rv_frequency` might work, especially if we were to accept non-normalized `pk` and normalize it (e.g. `pk /= np.sum(pk)`).

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distn_infrastructure.py:3946]: Yeah... let's leave that alone. I don't think we need to touch the tests. I am surprised at myself for even wanting to work on this, but gh-8057 has been nagging me.

**Comment by mdhaber** [comment]
> I'm not sure how you want it to play out, would you expand?
https://github.com/scipy/scipy/pull/17236#issuecomment-1279998536

So for instance, I think `pmf` currently has:
```python3
[x == k for k in self.xk]
```
I think that would just become:
```python3
[(x >= k_low) & (x <= k_high) for k in zip(self.xk_low, self.xk_high)]
```
or similar.

> Also whether you want it in the same PR

Well, I suppose I prefer separate PRs, but preferably an immediate followup. I wouldn't want to release this class before that is added.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/tests/test_discrete_distns.py:576]: I'd also show that you can free `s`.

**Comment by tupui** [comment]
Sounds like a good fix until we figure out the next API 👍 This could use some doc as well I think.

**Comment by ev-br** [comment]
OK, renamed to `rv_count`, as per https://github.com/scipy/scipy/pull/17236#discussion_r996653549, made other changes from review comments and added a docstring.

**Comment by ev-br** [comment]
Email sent: https://mail.python.org/archives/list/scipy-dev@python.org/thread/VLAMNUQFGRQ2U76DED2Z5RFG6Q4DWWJR/

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:3951]: Do we still need these (especially `longname`)? If so, can we accept them as `kwargs` and not document them?

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:1057]: I suppose we could leave the name of `rv_sample` alone now that the public class is `rv_count`. Not necessary to change back; just a thought.

**Review by mdhaber — APPROVED** [review]
Aside from the lint issue and two comments to think about (not necessarily implement), this LGTM!

File: scipy/stats/_distn_infrastructure.py
Comment (ev-br) [reply scipy/stats/_distn_infrastructure.py:3951]: A short `name` is somewhat useful I think. E.g. `rv_count(..., name='cat count')` and `rv_count(..., name='dog count')`.  `longname` may be indeed there for backwards compat only, so can in principle remove. Going through kwargs however adds possible user confusion (what other kwargs are they accepting and what they mean?) and a burden of validating them. So I'd rather keep this as is.

File: scipy/stats/_distn_infrastructure.py
Comment (ev-br) [reply scipy/stats/_distn_infrastructure.py]: Good points, `rv_sample` is back in the last commit.

**Comment by ev-br** [comment]
The lint "issue" is this line being too long:

```
    seed : {None, int, `numpy.random.Generator`, `numpy.random.RandomState`}, optional
```

The line is copy-pasted from docstrings of other distributions. I'd personally value consistency more than a line being a whopping 86 characters TBH.

**Comment by mdhaber** [comment]
I know we have it broken into two lines in some stats functions - but I also know that if it's done incorrectly it renders funny. So if it won't make CI complain in `main` after merge, I'll merge it like this after your email has been out for a week.

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:3952]: I just had another PR merge with a lint failure and I saw the issue in another PR. I don't want that to happen here, so I'm changing this to how it appears in some other stats functions.
```suggestion
    seed : {None, int, `numpy.random.Generator`,
            `numpy.random.RandomState`}, optional

```

**Comment by mdhaber** [comment]
There are a few things I'd like to adjust before this is released:

1.
```python3
dist = stats.rv_count([1, 1], [0.5, 0.5])
dist.pmf([0, 1, 2])  # array([0. , 0.5, 0. ])
```
We could raise an error or add up the probabilities in the bins. I think that latter could be useful, actually. I suggested "count" in the name because I thought that this might be useful to model observed discrete data. So if we add up all the probabilities of non-unique elements, and the user has all their distinct observations in `data`, `rv_count` could be instantiated like:
```python3
dist = rv_count(data, 1/len(data))
```
assuming `xk` and `pk` get broadcasted. Otherwise, they have to compute the proportions of all their non-unique elements themselves. 

2. 
```python3
dist = stats.rv_count([1, 2], [0.5, 0.5-1e-4])  # ValueError: The sum of provided pk is not 1.
dist = stats.rv_count([1, 2], [0.5, 0.5-0.999e-5])  # No problemo!
```
Why even bother with defining a cutoff and raising an error? Instead, if we just document that `pk` will be normalized, then there is no arbitrary distinction between these cases, we don't need to raise (and test) an error, and the user can now provide counts (as in `rv_count`) instead of probabilities.

3.
```python3
dist = stats.rv_count([1.1, 2], [1/2, 1/2])
dist_frozen = dist(loc = 0.8)
dist_frozen.pmf(1.9)  # 0 because:
1.1 + 0.8 == 1.9  # False
```
We expected this. But that's not the real reason _currently_; right now it's actually worse than that:

4.
```python3
dist = stats.rv_count([0.5, 1], [1/2, 1/2])
dist.pmf([0.5, 1])  # array([0. , 0.5])
```
This is because of gh-17166. This new class is no longer called `rv_sample`, so it's not excused from the integrality requirement.

5.
```python3
dist = stats.rv_count([0.5, 1.5], [1/2, 1/2])
dist.rvs(size=10)  # array([1, 1, 0, 0, 1, 1, 1, 1, 0, 0], dtype=int64)
```
I'm not sure what that's all about.

These were actually the first few things I tried when playing with this, so I'm guessing there are other issues. Of course, these are pre-existing conditions; they were not caused by PR. This PR just makes it easier for users to find them.

**Pull request merged**
